### PR TITLE
Consider `import.meta` a qualified name/property access

### LINKED ISSUE #54590: Completions request can trigger incorrect subsequent type-checking errors
In VS Code, do the following:

```ts
let x = import.meta[|/*hello*/|];
```

Notice that the entire range of `/*hello*/` is highlighted. It doesn't have to be a comment, it can be a range of whitespace trivia or something.

Delete the entire range. You will then get an error

```
Cannot find name 'meta'.
```

image

If you look at the following request, you can see that a `completionInfo`/`completionEntryDetails` request was triggered before `geterr`:

<details>
<summary>Output Pane Log</summary>

```
2023-06-09 12:31:33.376 [trace] <syntax> Sending request: updateOpen (225). Response expected: yes. Current queue length: 0 {
    "changedFiles": [
        {
            "fileName": "^/untitled/ts-nul-authority/Untitled-1",
            "textChanges": [
                {
                    "newText": "",
                    "start": {
                        "line": 1,
                        "offset": 20
                    },
                    "end": {
                        "line": 1,
                        "offset": 29
                    }
                }
            ]
        }
    ],
    "closedFiles": [],
    "openFiles": []
}
2023-06-09 12:31:33.376 [trace] <semantic> Sending request: updateOpen (267). Response expected: yes. Current queue length: 0 {
    "changedFiles": [
        {
            "fileName": "^/untitled/ts-nul-authority/Untitled-1",
            "textChanges": [
                {
                    "newText": "",
                    "start": {
                        "line": 1,
                        "offset": 20
                    },
                    "end": {
                        "line": 1,
                        "offset": 29
                    }
                }
            ]
        }
    ],
    "closedFiles": [],
    "openFiles": []
}
2023-06-09 12:31:33.377 [trace] <semantic> Response received: updateOpen (267). Request took 1 ms. Success: true  true
2023-06-09 12:31:33.377 [trace] <semantic> Sending request: completionInfo (268). Response expected: yes. Current queue length: 0 {
    "file": "^/untitled/ts-nul-authority/Untitled-1",
    "line": 1,
    "offset": 20,
    "includeExternalModuleExports": true,
    "includeInsertTextCompletions": true,
    "triggerKind": 1
}
2023-06-09 12:31:33.377 [trace] <syntax> Response received: updateOpen (225). Request took 2 ms. Success: true  true
2023-06-09 12:31:33.386 [trace] <semantic> Response received: completionInfo (268). Request took 11 ms. Success: true  {
    "flags": 0,
    "isGlobalCompletion": false,
    "isMemberCompletion": true,
    "isNewIdentifierLocation": false,
    "optionalReplacementSpan": {
        "start": {
            "line": 1,
            "offset": 16
        },
        "end": {
            "line": 1,
            "offset": 20
        }
    },
    "entries": [
        {
            "name": "meta",
            "kind": "property",
            "kindModifiers": "",
            "sortText": "11"
        }
    ]
}
2023-06-09 12:31:33.393 [trace] <semantic> Sending request: completionEntryDetails (269). Response expected: yes. Current queue length: 0 {
    "file": "^/untitled/ts-nul-authority/Untitled-1",
    "line": 1,
    "offset": 20,
    "entryNames": [
        "meta"
    ]
}
2023-06-09 12:31:33.396 [trace] <semantic> Response received: completionEntryDetails (269). Request took 4 ms. Success: true  [
    {
        "name": "meta",
        "kindModifiers": "",
        "kind": "property",
        "displayParts": [
            {
                "text": "(",
                "kind": "punctuation"
            },
            {
                "text": "property",
                "kind": "text"
            },
            {
                "text": ")",
                "kind": "punctuation"
            },
            {
                "text": " ",
                "kind": "space"
            },
            {
                "text": "ImportMetaExpression",
                "kind": "text"
            },
            {
                "text": ".",
                "kind": "punctuation"
            },
            {
                "text": "meta",
                "kind": "propertyName"
            },
            {
                "text": ":",
                "kind": "punctuation"
            },
            {
                "text": " ",
                "kind": "space"
            },
            {
                "text": "ImportMeta",
                "kind": "interfaceName"
            }
        ],
        "documentation": [],
        "tags": []
    }
]
2023-06-09 12:31:33.421 [trace] <syntax> Sending request: linkedEditingRange (226). Response expected: yes. Current queue length: 0 {
    "file": "^/untitled/ts-nul-authority/Untitled-1",
    "line": 1,
    "offset": 20
}
2023-06-09 12:31:33.422 [trace] <syntax> Response received: linkedEditingRange (226). Request took 2 ms. Success: false . Message: No content available.
2023-06-09 12:31:33.557 [trace] <syntax> Sending request: getOutliningSpans (227). Response expected: yes. Current queue length: 0 {
    "file": "^/untitled/ts-nul-authority/Untitled-1"
}
2023-06-09 12:31:33.558 [trace] <syntax> Response received: getOutliningSpans (227). Request took 1 ms. Success: true  []
2023-06-09 12:31:33.612 [trace] <semantic> Sending request: getApplicableRefactors (270). Response expected: yes. Current queue length: 0 {
    "file": "^/untitled/ts-nul-authority/Untitled-1",
    "startLine": 1,
    "startOffset": 20,
    "endLine": 1,
    "endOffset": 20,
    "includeInteractiveActions": false
}
2023-06-09 12:31:33.647 [trace] <semantic> Response received: getApplicableRefactors (270). Request took 35 ms. Success: true  [
    {
        "name": "Convert export",
        "description": "Convert default export to named export",
        "actions": [
            {
                "name": "Convert default export to named export",
                "description": "Convert default export to named export",
                "kind": "refactor.rewrite.export.named",
                "notApplicableReason": "Could not find export statement"
            },
            {
                "name": "Convert named export to default export",
                "description": "Convert named export to default export",
                "kind": "refactor.rewrite.export.default",
                "notApplicableReason": "Could not find export statement"
            }
        ]
    },
    {
        "name": "Convert import",
        "description": "Convert namespace import to named imports",
        "actions": [
            {
                "name": "Convert namespace import to named imports",
                "description": "Convert namespace import to named imports",
                "kind": "refactor.rewrite.import.named",
                "notApplicableReason": "Selection is not an import declaration."
            }
        ]
    },
    {
        "name": "Convert import",
        "description": "Convert named imports to default import",
        "actions": [
            {
                "name": "Convert named imports to default import",
                "description": "Convert named imports to default import",
                "kind": "refactor.rewrite.import.default",
                "notApplicableReason": "Selection is not an import declaration."
            }
        ]
    },
    {
        "name": "Convert import",
        "description": "Convert named imports to namespace import",
        "actions": [
            {
                "name": "Convert named imports to namespace import",
                "description": "Convert named imports to namespace import",
                "kind": "refactor.rewrite.import.namespace",
                "notApplicableReason": "Selection is not an import declaration."
            }
        ]
    },
    {
        "name": "Extract type",
        "description": "Extract type",
        "actions": [
            {
                "name": "Extract to typedef",
                "description": "Extract to typedef",
                "kind": "refactor.extract.typedef",
                "notApplicableReason": "Selection is not a valid type node"
            },
            {
                "name": "Extract to type alias",
                "description": "Extract to type alias",
                "kind": "refactor.extract.type",
                "notApplicableReason": "Selection is not a valid type node"
            },
            {
                "name": "Extract to interface",
                "description": "Extract to interface",
                "kind": "refactor.extract.interface",
                "notApplicableReason": "Selection is not a valid type node"
            }
        ]
    },
    {
        "name": "Move to a new file",
        "description": "Move to a new file",
        "actions": [
            {
                "name": "Move to a new file",
                "description": "Move to a new file",
                "kind": "refactor.move.newFile",
                "notApplicableReason": "Selection is not a valid statement or statements"
            }
        ]
    },
    {
        "name": "Add or remove braces in an arrow function",
        "description": "Add or remove braces in an arrow function",
        "actions": [
            {
                "name": "Add braces to arrow function",
                "description": "Add braces to arrow function",
                "kind": "refactor.rewrite.arrow.braces.add",
                "notApplicableReason": "Could not find a containing arrow function"
            },
            {
                "name": "Remove braces from arrow function",
                "description": "Remove braces from arrow function",
                "kind": "refactor.rewrite.arrow.braces.remove",
                "notApplicableReason": "Could not find a containing arrow function"
            }
        ]
    },
    {
        "name": "Convert to template string",
        "description": "Convert to template string",
        "actions": [
            {
                "name": "Convert to template string",
                "description": "Convert to template string",
                "kind": "refactor.rewrite.string",
                "notApplicableReason": "Can only convert string concatenation"
            }
        ]
    },
    {
        "name": "Extract Symbol",
        "description": "Extract function",
        "actions": [
            {
                "name": "Extract Function",
                "description": "Extract function",
                "kind": "refactor.extract.function",
                "notApplicableReason": "Cannot extract empty range."
            }
        ]
    },
    {
        "name": "Extract Symbol",
        "description": "Extract constant",
        "actions": [
            {
                "name": "Extract Constant",
                "description": "Extract constant",
                "kind": "refactor.extract.constant",
                "notApplicableReason": "Cannot extract empty range."
            }
        ]
    },
    {
        "name": "Generate 'get' and 'set' accessors",
        "description": "Generate 'get' and 'set' accessors",
        "actions": [
            {
                "name": "Generate 'get' and 'set' accessors",
                "description": "Generate 'get' and 'set' accessors",
                "kind": "refactor.rewrite.property.generateAccessors",
                "notApplicableReason": "Could not find property for which to generate accessor"
            }
        ]
    },
    {
        "name": "Infer function return type",
        "description": "Infer function return type",
        "actions": [
            {
                "name": "Infer function return type",
                "description": "Infer function return type",
                "kind": "refactor.rewrite.function.returnType",
                "notApplicableReason": "Return type must be inferred from a function"
            }
        ]
    }
]
2023-06-09 12:31:33.664 [trace] <semantic> Sending request: geterr (271). Response expected: yes. Current queue length: 0 {
    "delay": 0,
    "files": [
        "^/untitled/ts-nul-authority/Untitled-1"
    ]
}
2023-06-09 12:31:33.680 [trace] <semantic> Event received: syntaxDiag (0). {
    "file": "^/untitled/ts-nul-authority/Untitled-1",
    "diagnostics": []
}
2023-06-09 12:31:33.682 [trace] <semantic> Event received: semanticDiag (0). {
    "file": "^/untitled/ts-nul-authority/Untitled-1",
    "diagnostics": [
        {
            "start": {
                "line": 1,
                "offset": 16
            },
            "end": {
                "line": 1,
                "offset": 20
            },
            "text": "Cannot find name 'meta'.",
            "code": 2304,
            "category": "error"
        }
    ]
}
2023-06-09 12:31:33.682 [trace] <semantic> Event received: suggestionDiag (0). {
    "file": "^/untitled/ts-nul-authority/Untitled-1",
    "diagnostics": [
        {
            "start": {
                "line": 1,
                "offset": 5
            },
            "end": {
                "line": 1,
                "offset": 6
            },
            "text": "'x' is declared but its value is never read.",
            "code": 6133,
            "category": "suggestion",
            "reportsUnnecessary": true
        }
    ]
}
2023-06-09 12:31:33.683 [trace] <semantic> Async response received: requestCompleted (271). Request took 19 ms.
```

</details>

You can *probably* repro this with something like the following fourslash test, though I haven't tried:

```ts
/// <reference path="fourslash.ts" />

//// let x = import.meta/**/;

verify./*someCompletionsRequest*;
verify./*someErrorsRequest*/;
```

### PR DESCRIPTION
**Pull request by gabritto**
Fixes #54590.

`isRightSideOfQualifiedNameOrPropertyAccess` is called by `getTypeOfSymbolAtLocation` to decide if we should consider the location's parent instead, so that if we have a property access/qualified name, e.g. `x.y`, we look up the whole expression, and not just `y`. Because `import.meta` was not considered one of those cases, when `getTypeOfSymbolAtLocation` was called by completions, we'd end up trying to resolve the name `meta`, which doesn't resolve to anything, leading to the error.

### CODE REVIEW

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot run DT
@typescript-bot user test tsserver
@typescript-bot test top100
@typescript-bot test tsserver top100
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot test this
@typescript-bot run DT
@typescript-bot user test tsserver
@typescript-bot test top100
@typescript-bot test tsserver top100
@typescript-bot perf test this

File: tests/baselines/reference/plainJSGrammarErrors.types [hunk 1]
```diff
@@ -606,7 +606,7 @@ export let noMeta = import.metal
 function foo() { new.targe }
 >foo : () => void
 >new.targe : () => void
->targe : any
+>targe : () => void
```
Comment (jakebailey) [inline tests/baselines/reference/plainJSGrammarErrors.types:609]: Weird, how is this changing? This doesn't seem to be using meta at all.

Comment (gabritto) [reply tests/baselines/reference/plainJSGrammarErrors.types:609]: `new.something` is also a `MetaProperty` node, I'm guessing this is to also support `new.target`.

**Comment by gabritto** [comment]
@typescript-bot run DT

File: tests/baselines/reference/plainJSGrammarErrors.types (hunk 1 above)
Comment (DanielRosenwasser) [reply tests/baselines/reference/plainJSGrammarErrors.types:609]: Probably fall-out logic where any meta-property that's not `import.meta` is assumed to be `new.target`.

**Pull request merged**